packages/core/src/typeschema/validation.ts

```diff
@@ -580,7 +580,12 @@ export function matchDiscriminant(
     return false;
   }
 
-  const sliceElement: InternalSchemaElement = (elements ?? slice.elements)[discriminator.path];
+  let sliceElement: InternalSchemaElement | undefined;
+  if (discriminator.path === '$this') {
```

Comment:
Clarifying question:  My understanding is that `discriminator.path` can be a full FHIRPath expression, so we're currently adding support for the special case of `=== '$this'`, but still have open future work here?

Reply:
Yes, that's correct; the [Restricted Subset ("Simple")](https://hl7.org/fhir/R4/fhirpath.html#simple):

* All statements SHALL start with the name of the context element (e.g. on a Patient resource, Patient.contact.name.), or SHALL be simply "$this" to refer to the element that has focus
* Operators SHALL NOT be used
* Only the following functions may be used:
  * .resolve()
  * .extension("url")
  * .ofType(type)
* All other functions SHALL NOT be used